### PR TITLE
Main page

### PR DESCRIPTION
**Pull request by dsekulovic**

### CODE REVIEW

File: src/app/app.component.html [hunk 1]
```diff
@@ -1,7 +1,25 @@
 <div class="content" role="main">
```
Comment (BraneOpacic21) [inline src/app/app.component.html]: We should not have empty rows in our HTML/template - Please add some code formatter (prettier, beautify etc.).

File: src/app/app.component.html [hunk 2]
```diff
@@ -1,7 +1,25 @@
...
+        </li>
+        <li class="nav-item" routerLinkActive="active">
+          <a class="nav-link" routerLink="/location">Countries</a>
+        </li>
+        <li class="nav-item" routerLinkActive="active">
+          <a class="nav-link" routerLink="/about">About Covid-19</a>
+        </li>
+      </ul>
+    </div>
+  </nav>
```
Comment (BraneOpacic21) [inline src/app/app.component.html]: Same as above

File: src/app/app.component.html [hunk 3]
```diff
@@ -1,7 +1,25 @@
...
 
-</div>
+  <router-outlet></router-outlet>
 
-<router-outlet></router-outlet>
\ No newline at end of file
+</div>
```
Comment (BraneOpacic21) [inline src/app/app.component.html:22]: /n

File: src/app/app.module.ts [hunk 4]
```diff
@@ -1,18 +1,38 @@
-import { BrowserModule } from '@angular/platform-browser';
-import { NgModule } from '@angular/core';
+import { BrowserModule } from "@angular/platform-browser";
+import { NgModule } from "@angular/core";
+import { HttpClientModule } from "@angular/common/http";
+import { ChartsModule } from "ng2-charts";
+import { AgmCoreModule } from "@agm/core";
```
Comment (BraneOpacic21) [inline src/app/app.module.ts]: remove empty row

File: src/app/app.module.ts [hunk 5]
```diff
@@ -1,18 +1,38 @@
...
 
-import { AppRoutingModule } from './app-routing.module';
-import { AppComponent } from './app.component';
+import { AppRoutingModule } from "./app-routing.module";
+import { GOOGLE_API_KEY } from "./utility/constants";
+
```
Comment (BraneOpacic21) [inline src/app/app.module.ts]: remove empty row

File: src/app/charts/charts.component.html [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+        <label class="btn" [ngClass]="{
+            'btn-danger active':label === 'TotalConfirmed' || label === 'NewConfirmed',
+            'btn-success':label === 'TotalRecovered' || label === 'NewRecovered',
+            'btn-dark':label === 'TotalDeaths' || label === 'NewDeaths'
+          }" *ngFor="let label of labels">
+            <input type="radio" name="options" (click)="onClick(label)">
+            {{ label }}
+        </label>
+    </div>
+
```
Comment (BraneOpacic21) [inline src/app/charts/charts.component.html]: remove row

File: src/app/charts/charts.component.html [hunk 7]
```diff
@@ -0,0 +1,19 @@
+<div class="mt-3 justify-content-center">
+    <div class="btn-group btn-group-sm btn-group-toggle" data-toggle="buttons">
+        <label class="btn" [ngClass]="{
+            'btn-danger active':label === 'TotalConfirmed' || label === 'NewConfirmed',
+            'btn-success':label === 'TotalRecovered' || label === 'NewRecovered',
+            'btn-dark':label === 'TotalDeaths' || label === 'NewDeaths'
+          }" *ngFor="let label of labels">
+            <input type="radio" name="options" (click)="onClick(label)">
```
Comment (BraneOpacic21) [inline src/app/charts/charts.component.html:8]: We should somehow hook the label with the input

File: src/app/charts/charts.component.html [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+        </label>
+    </div>
+
+    <div class="mt-4">
+        <canvas height="200" baseChart [datasets]="lineChartData" [labels]="lineChartLabels" [chartType]="lineChartType"
+            [colors]="lineChartColors">
+        </canvas>
+        <h5 class="text-center py-2">{{numberOfData}} countries with most {{activeSort}} cases</h5>
+    </div>
+</div>
```
Comment (BraneOpacic21) [inline src/app/charts/charts.component.html]: /n

File: src/app/charts/charts.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (BraneOpacic21) [inline src/app/charts/charts.component.spec.ts]: No need for testing in this app - you can remove all the test files 🕺

File: src/app/charts/charts.component.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    this.lineChartLabels = [];
+    this.lineChartData = [];
+    this.loadChartData(type);
+  }
+
+  loadChartData(type: string) {
+    this.activeSort = type;
+
+    for (const label of this.labels) {
+      this.lineChartData.push({ data: [], label });
```
Comment (BraneOpacic21) [inline src/app/charts/charts.component.ts]: Could we use Array.map here? It would be more readable, and also - it would return a new array, preventing immutability problems

File: src/app/charts/charts.component.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    this.chartData
+      .sort((a: CountryInfo, b: CountryInfo) => (a[type] < b[type] ? 1 : -1))
+      .slice(0, this.numberOfData)
+      .forEach((element: CountryInfo) => {
+        for (const bar of this.lineChartData) {
+          bar.data.push(element[bar.label]);
+        }
+
+        this.lineChartLabels.push(element.Country);
+      });
```
Comment (BraneOpacic21) [inline src/app/charts/charts.component.ts]: This should be done simpler. Hint: avoid using `for`

File: src/app/home/home.component.html [hunk 12]
```diff
@@ -0,0 +1,26 @@
+<div class="d-flex justify-content-center mt-5" *ngIf="isLoading && !error">
+    <div class="spinner-border" role="status">
+        <span class="sr-only">Loading...</span>
+    </div>
+</div>
+
```
Comment (BraneOpacic21) [inline src/app/home/home.component.html]: remove (to avoid repeating myself, it refers to all the template files)

File: src/app/home/home.component.html [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+        <app-charts class="col-lg-6" [chartData]="chartData"
+            [labels]="['TotalConfirmed','TotalRecovered','TotalDeaths']" [numberOfData]="20">
+        </app-charts>
+
+        <app-charts class="col-lg-6" [chartData]="chartData" [labels]="['NewConfirmed','NewRecovered','NewDeaths']"
+            [numberOfData]="20">
+        </app-charts>
+    </div>
+
+</div>
```
Comment (BraneOpacic21) [inline src/app/home/home.component.html]: /n

File: src/app/home/home.component.html [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    <p>{{ error }}</p>
+</div>
+
+<div class="container" *ngIf="!isLoading && !error">
+
+    <app-info-card class="row text-center mt-4" [data]="data"></app-info-card>
+
+    <div class="row">
+        <app-charts class="col-lg-6" [chartData]="chartData"
+            [labels]="['TotalConfirmed','TotalRecovered','TotalDeaths']" [numberOfData]="20">
```
Comment (BraneOpacic21) [inline src/app/home/home.component.html]: [labels] should probably be declared in a component (not a huge issue, just a better practice)

File: src/app/home/home.component.html [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+        </app-charts>
+
+        <app-charts class="col-lg-6" [chartData]="chartData" [labels]="['NewConfirmed','NewRecovered','NewDeaths']"
```
Comment (BraneOpacic21) [inline src/app/home/home.component.html]: same as above

File: src/app/home/home.component.ts [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+export class HomeComponent implements OnInit {
+  isLoading: boolean = false;
+  data: CovidInfo;
+  chartData: CountryInfo[];
+  error = null;
+
+  constructor(private http: HttpClass) {}
+
+  ngOnInit(): void {
+    this.isLoading = true;
```
Comment (BraneOpacic21) [inline src/app/home/home.component.ts]: Loading should be coupled with the async calls (fetchingData, e.g.), if possible.

File: src/app/home/home.component.ts [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+    this.fetchingData();
+  }
+
+  fetchingData() {
+    this.http.getTotalData().subscribe(
```
Comment (BraneOpacic21) [inline src/app/home/home.component.ts]: Are we unsubscribing from our subscriptions?

File: src/app/home/home.component.ts [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+      (data) => {
```
Comment (BraneOpacic21) [inline src/app/home/home.component.ts]: ({countries, global})

File: src/app/info-card/info-card.component.html [hunk 19]
```diff
@@ -0,0 +1,27 @@
+<div class="col-sm-4 mb-2">
+    <div class="card">
+        <div class="card-body">
+            <h5 class="card-title">Total confirmed:</h5>
+            <h3 class="card-text">{{data.TotalConfirmed }}</h3>
+            <p class="card-text text-danger" *ngIf="data.NewConfirmed">+{{data.NewConfirmed }}</p>
```
Comment (BraneOpacic21) [inline src/app/info-card/info-card.component.html]: This probably is not a `<p>`

File: src/app/info-card/info-card.component.html [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+        </div>
+    </div>
+</div>
+<div class="col-sm-4 mb-2">
+    <div class="card">
+        <div class="card-body">
+            <h5 class="card-title">Total recovered:</h5>
+            <h3 class="card-text">{{ data.TotalRecovered }}</h3>
+            <p class="card-text text-danger" *ngIf="data.NewRecovered">+{{ data.NewRecovered }}</p>
```
Comment (BraneOpacic21) [inline src/app/info-card/info-card.component.html]: same as above

File: src/app/info-card/info-card.component.html [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+        </div>
+    </div>
+</div>
+<div class="col-sm-4 mb-2">
+    <div class="card">
+        <div class="card-body">
+            <h5 class="card-title">Total deaths:</h5>
+            <h3 class="card-text">{{ data.TotalDeaths }}</h3>
+            <p class="card-text text-danger" *ngIf="data.NewDeaths">+{{ data.NewDeaths }}</p>
```
Comment (BraneOpacic21) [inline src/app/info-card/info-card.component.html]: same

File: src/app/info-card/info-card.component.html [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+        </div>
+    </div>
+</div>
```
Comment (BraneOpacic21) [inline src/app/info-card/info-card.component.html]: /n

File: src/app/info-card/info-card.component.ts [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+  selector: "app-info-card",
+  templateUrl: "./info-card.component.html",
+  styleUrls: ["./info-card.component.scss"],
+})
+export class InfoCardComponent implements OnInit {
+  @Input() data: CovidInfo;
+
+  constructor() {}
+
+  ngOnInit(): void {}
```
Comment (BraneOpacic21) [inline src/app/info-card/info-card.component.ts]: If we aren't using ngOnInit life-cycle hook - please remove it

File: src/app/interface/interface.ts [hunk 24]
```diff
@@ -0,0 +1,21 @@
+export interface CovidInfo {
```
Comment (BraneOpacic21) [inline src/app/interface/interface.ts]: I like this - very good. Keep the flow with the mandatory properties above the optional ones 👍 

FYI, good practice for naming interfaces is to put the capital I in front of the name (like: ICovidInfo)

File: src/index.html [hunk 25]
```diff
@@ -7,6 +7,10 @@
   <base href="/">
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <link rel="icon" type="image/x-icon" href="favicon.ico">
+  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.4.1/css/bootstrap.min.css">
+  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
```
Comment (BraneOpacic21) [inline src/index.html:11]: Do we need jQuery for something inside the app? 🤔

**Review by BraneOpacic21 — CHANGES_REQUESTED** [review]
Nice, in general 👍  Most of the comments are me repeating myself 😄

File: src/index.html (hunk 25 above)
Comment (dsekulovic) [reply src/index.html:11]: Neded for hamburger button to be toggled in navigation

File: src/app/charts/charts.component.html (hunk 7 above)
Comment (dsekulovic) [reply src/app/charts/charts.component.html:8]: The input is inside the label, or you ment something else?

File: src/index.html (hunk 25 above)
Comment (BraneOpacic21) [reply src/index.html:11]: Ok 👍

File: src/app/charts/charts.component.html (hunk 7 above)
Comment (BraneOpacic21) [reply src/app/charts/charts.component.html:8]: Ok, leave it as is!

**Review by BraneOpacic21 — APPROVED** [review]
For future reference: keep in mind that naming convention for subscriptions is with `$` at the end of the name, like so:

myVariable$
